### PR TITLE
JAMES-2706 Normalize metric contract testing

### PR DESCRIPTION
**Pull request by hoangdat**

### CODE REVIEW

File: metrics/metrics-tests/pom.xml [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+    <name>Apache James :: Metrics :: Tests</name>
+
+    <dependencies>
+        <dependency>
+            <groupId>com.google.guava</groupId>
+            <artifactId>guava</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>org.apache.james</groupId>
```
Comment (trantienduchn) [inline metrics/metrics-tests/pom.xml]: use ${james.groupId}

File: metrics/metrics-tests/src/main/java/org/apache/james/metrics/tests/RecordingMetricFactory.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+import org.apache.commons.lang3.NotImplementedException;
+import org.apache.james.metrics.api.Metric;
+import org.apache.james.metrics.api.MetricFactory;
+import org.apache.james.metrics.api.TimeMetric;
+
+import com.google.common.collect.HashMultimap;
+import com.google.common.collect.Multimap;
+import com.google.common.collect.Multimaps;
+
+public class RecordingMetricFactory implements MetricFactory {
```
Comment (trantienduchn) [inline metrics/metrics-tests/src/main/java/org/apache/james/metrics/tests/RecordingMetricFactory.java]: I think those classes are for using in tests, then they should be placed in `src/test` instead of `src/main`

File: server/blob/blob-api/src/test/java/org/apache/james/blob/api/MetricableBlobStoreContract.java [hunk 3]
```diff
@@ -79,32 +57,43 @@ private void setupExpectations() {
...
 
     @Test
     default void saveInputStreamShouldPublishSaveInputStreamTimerMetrics() {
+        long sleepDurationInMs = Duration.ofSeconds(1).toMillis();
+
         testee().save(new ByteArrayInputStream(BYTES_CONTENT), BYTES_CONTENT.length).block();
-        testee().save(new ByteArrayInputStream(BYTES_CONTENT), BYTES_CONTENT.length).block();
-        testee().save(new ByteArrayInputStream(BYTES_CONTENT), BYTES_CONTENT.length).block();
-        verify(metricsTestExtension.saveInputStreamTimeMetric, times(3)).stopAndPublish();
+
```
Comment (trantienduchn) [inline server/blob/blob-api/src/test/java/org/apache/james/blob/api/MetricableBlobStoreContract.java]: why did you remove those testee calls? I find it's helpful to know that metric recording is working well with multiple calls

File: metrics/metrics-tests/src/main/java/org/apache/james/metrics/tests/RecordingMetricFactory.java (hunk 2 above)
Comment (hoangdat) [reply metrics/metrics-tests/src/main/java/org/apache/james/metrics/tests/RecordingMetricFactory.java]: complete agree. But I can not make it available for other test modules

Comment (trantienduchn) [reply metrics/metrics-tests/src/main/java/org/apache/james/metrics/tests/RecordingMetricFactory.java]: importing test-jar ?

Comment (chibenwa) [reply metrics/metrics-tests/src/main/java/org/apache/james/metrics/tests/RecordingMetricFactory.java]: nope importing test jar is a hell.

When that is code that is reused somewhere else, declaring it in src/main is better

**Comment by chibenwa** [comment]
No JIRA ticket?

File: server/blob/blob-api/src/test/java/org/apache/james/blob/api/MetricableBlobStoreContract.java [hunk 4]
```diff
@@ -79,32 +57,43 @@ private void setupExpectations() {
 
     @Test
     default void saveBytesShouldPublishSaveBytesTimerMetrics() {
+        long sleepDurationInMs = Duration.ofSeconds(1).toMillis();
```
Comment (chibenwa) [inline server/blob/blob-api/src/test/java/org/apache/james/blob/api/MetricableBlobStoreContract.java]: Can we get smaller sleeps? 100 ms should be enough but will be fast enough to not annoy us IMO

**Comment by chibenwa** [comment]
See DeliveryRunnableTest LocalDeliveryTest MailQueueMetricExtension MetricableBlobStoreContract MetricsMailetTest ToRecipientFolderTest for many more tests using mock metrics!

**Comment by Arsnael** [comment]
> No JIRA ticket?

https://issues.apache.org/jira/browse/JAMES-2706

**Comment by chibenwa** [comment]
>> No JIRA ticket?

> https://issues.apache.org/jira/browse/JAMES-2706

You need to add it to the commit message

File: metrics/metrics-tests/src/main/java/org/apache/james/metrics/tests/RecordingTimeMetric.java [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+import java.util.function.Consumer;
+
+import org.apache.james.metrics.api.TimeMetric;
+
+import com.google.common.base.Stopwatch;
+
+public class RecordingTimeMetric implements TimeMetric {
+    private final String name;
+    private final Stopwatch stopwatch = Stopwatch.createStarted();
+    private final Consumer<Long> publishCallback;
```
Comment (chibenwa) [inline metrics/metrics-tests/src/main/java/org/apache/james/metrics/tests/RecordingTimeMetric.java]: Is that long a duration?

File: metrics/metrics-tests/src/main/java/org/apache/james/metrics/tests/RecordingMetricFactory.java [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+    private final Multimap<String, Long> executionTimesInMs = Multimaps.synchronizedSetMultimap(HashMultimap.create());
```
Comment (chibenwa) [inline metrics/metrics-tests/src/main/java/org/apache/james/metrics/tests/RecordingMetricFactory.java]: Long here is a duration...

File: metrics/metrics-tests/src/main/java/org/apache/james/metrics/tests/RecordingMetricFactory.java [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+
+import org.apache.james.metrics.api.Metric;
+import org.apache.james.metrics.api.MetricFactory;
+import org.apache.james.metrics.api.TimeMetric;
+
+import com.google.common.collect.HashMultimap;
+import com.google.common.collect.Multimap;
+import com.google.common.collect.Multimaps;
+
+public class RecordingMetricFactory implements MetricFactory {
```
Comment (chibenwa) [inline metrics/metrics-tests/src/main/java/org/apache/james/metrics/tests/RecordingMetricFactory.java]: We need a little test class now that we rely on this method in many places.

File: metrics/metrics-tests/src/main/java/org/apache/james/metrics/tests/RecordingMetricFactory.java [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+
+    public Collection<Long> executionTimesFor(String name) {
+        synchronized (executionTimesInMs) {
+            return executionTimesInMs.get(name);
+        }
+    }
+
+    public int executionValuesFor(String name) {
+        Integer value = executionValues.get(name);
+        if (value == null) {
```
Comment (chibenwa) [inline metrics/metrics-tests/src/main/java/org/apache/james/metrics/tests/RecordingMetricFactory.java]: is that a disguised `getOrDefault`?

File: metrics/metrics-tests/src/main/java/org/apache/james/metrics/tests/RecordingMetricFactory.java [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+        });
+    }
+
+    public Collection<Long> executionTimesFor(String name) {
+        synchronized (executionTimesInMs) {
+            return executionTimesInMs.get(name);
+        }
+    }
+
+    public int executionValuesFor(String name) {
```
Comment (chibenwa) [inline metrics/metrics-tests/src/main/java/org/apache/james/metrics/tests/RecordingMetricFactory.java]: s/executionValuesFor/countFor/

File: server/blob/blob-api/src/test/java/org/apache/james/blob/api/MetricableBlobStoreContract.java [hunk 10]
```diff
@@ -79,32 +57,51 @@ private void setupExpectations() {
...
+        long durationInMs = Duration.ofMillis(100).toMillis();
+
         BlobId blobId = testee().save(BYTES_CONTENT).block();
         testee().read(blobId);
         testee().read(blobId);
-        verify(metricsTestExtension.readTimeMetric, times(2)).stopAndPublish();
+
+        assertThat(metricsTestExtension.getMetricFactory().executionTimesFor(READ_TIMER_NAME))
+            .hasSize(2)
+            .allSatisfy(executionInMs -> assertThat(executionInMs).isLessThan(durationInMs));
```
Comment (chibenwa) [inline server/blob/blob-api/src/test/java/org/apache/james/blob/api/MetricableBlobStoreContract.java]: This allStatify do not add value

File: server/blob/blob-api/src/test/java/org/apache/james/blob/api/MetricableBlobStoreContract.java [hunk 11]
```diff
@@ -79,32 +57,51 @@ private void setupExpectations() {
...
+        long durationInMs = Duration.ofMillis(100).toMillis();
+
         BlobId blobId = testee().save(BYTES_CONTENT).block();
         testee().readBytes(blobId).block();
         testee().readBytes(blobId).block();
-        verify(metricsTestExtension.readBytesTimeMetric, times(2)).stopAndPublish();
+
+        assertThat(metricsTestExtension.getMetricFactory().executionTimesFor(READ_BYTES_TIMER_NAME))
+            .hasSize(2)
+            .allSatisfy(executionInMs -> assertThat(executionInMs).isLessThan(durationInMs));
```
Comment (chibenwa) [inline server/blob/blob-api/src/test/java/org/apache/james/blob/api/MetricableBlobStoreContract.java]: This allStatify do not add value

File: server/blob/blob-api/src/test/java/org/apache/james/blob/api/MetricableBlobStoreContract.java [hunk 12]
```diff
@@ -79,32 +57,51 @@ private void setupExpectations() {
...
+        long durationInMs = Duration.ofMillis(100).toMillis();
+
         testee().save(new ByteArrayInputStream(BYTES_CONTENT), BYTES_CONTENT.length).block();
         testee().save(new ByteArrayInputStream(BYTES_CONTENT), BYTES_CONTENT.length).block();
-        testee().save(new ByteArrayInputStream(BYTES_CONTENT), BYTES_CONTENT.length).block();
-        verify(metricsTestExtension.saveInputStreamTimeMetric, times(3)).stopAndPublish();
+
+        assertThat(metricsTestExtension.getMetricFactory().executionTimesFor(SAVE_INPUT_STREAM_TIMER_NAME))
+            .hasSize(2)
+            .allSatisfy(executionInMs -> assertThat(executionInMs).isLessThan(durationInMs));
```
Comment (chibenwa) [inline server/blob/blob-api/src/test/java/org/apache/james/blob/api/MetricableBlobStoreContract.java]: This allStatify do not add value

File: server/blob/blob-api/src/test/java/org/apache/james/blob/api/MetricableBlobStoreContract.java [hunk 13]
```diff
@@ -79,32 +57,51 @@ private void setupExpectations() {
...
     default void saveBytesShouldPublishSaveBytesTimerMetrics() {
+        long durationInMs = Duration.ofMillis(100).toMillis();
+
         testee().save(BYTES_CONTENT).block();
         testee().save(BYTES_CONTENT).block();
-        verify(metricsTestExtension.saveBytesTimeMetric, times(2)).stopAndPublish();
+
+        assertThat(metricsTestExtension.getMetricFactory().executionTimesFor(SAVE_BYTES_TIMER_NAME))
+            .hasSize(2)
+            .allSatisfy(executionInMs -> assertThat(executionInMs).isLessThan(durationInMs));
```
Comment (chibenwa) [inline server/blob/blob-api/src/test/java/org/apache/james/blob/api/MetricableBlobStoreContract.java]: This allStatify do not add value

File: server/queue/queue-activemq/src/test/java/org/apache/james/queue/activemq/ActiveMQMailQueueBlobTest.java [hunk 14]
```diff
@@ -95,6 +94,11 @@ public ManageableMailQueue getManageableMailQueue() {
         return mailQueue;
     }
 
+    @Override
+    public String queueName() {
```
Comment (chibenwa) [inline server/queue/queue-activemq/src/test/java/org/apache/james/queue/activemq/ActiveMQMailQueueBlobTest.java]: We should find a way to not need this IMO - I bet there is only one queue in the given recording metric, right? maybe we could expose a copy of its internal map?

File: server/queue/queue-api/src/test/java/org/apache/james/queue/api/MailQueueMetricContract.java [hunk 15]
```diff
@@ -74,51 +88,57 @@ default void constructorShouldRegisterGetQueueSizeGauge(MailQueueMetricExtension
...
     default void enqueueShouldPublishEnqueueTimeMetric(MailQueueMetricExtension.MailQueueMetricTestSystem testSystem) throws Exception {
+        long durationInMs = java.time.Duration.ofSeconds(1).toMillis();
+
         enQueueMail(2);
 
-        verify(testSystem.getSpyEnqueuedMailsTimeMetric(), times(2)).stopAndPublish();
+        WAIT_CONDITION
+            .until(() -> assertThat(testSystem.getMetricFactory().executionTimesFor(getMetricName(ENQUEUED_TIMER_METRIC_NAME_PREFIX)))
+                .hasSize(2)
+                .allSatisfy(executionInMs -> assertThat(executionInMs).isLessThan(durationInMs)));
```
Comment (chibenwa) [inline server/queue/queue-api/src/test/java/org/apache/james/queue/api/MailQueueMetricContract.java]: IMO this allSatisfy do not add value

File: server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/remote/delivery/DeliveryRunnable.java [hunk 16]
```diff
@@ -47,7 +48,7 @@
 
     public static final Supplier<Date> CURRENT_DATE_SUPPLIER = Date::new;
     public static final AtomicBoolean DEFAULT_NOT_STARTED = new AtomicBoolean(false);
-    private static final String OUTGOING_MAILS = "outgoingMails";
+    public static final String OUTGOING_MAILS = "outgoingMails";
```
Comment (chibenwa) [inline server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/remote/delivery/DeliveryRunnable.java]: Is package private enough?

File: server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/MetricsMailetTest.java [hunk 17]
```diff
@@ -69,8 +60,8 @@ public void serviceShouldIncrementMetricCounter() throws Exception {
 
         mailet.service(FakeMail.builder().name("name").build());
 
-        verify(metric).increment();
-        verifyNoMoreInteractions(metric);
+        assertThat(metricFactory.executionValuesFor("metricName"))
```
Comment (chibenwa) [inline server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/MetricsMailetTest.java]: Extract `metricName` to a constant

File: server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/delivery/LocalDeliveryTest.java [hunk 18]
```diff
@@ -121,6 +121,49 @@ public void mailShouldBeWellDeliveredByDefaultToUserWhenVirtualHostingIsTurnedOf
...
+    public void mailDeliveredShouldIncrementMetricCounter() throws Exception {
+        // Given
+        String username = "receiver";
+        MailboxPath inbox = MailboxPath.forUser(username, "INBOX");
+        MessageManager messageManager = mock(MessageManager.class);
+        when(usersRepository.supportVirtualHosting()).thenReturn(false);
+        when(usersRepository.getUser(new MailAddress("receiver@localhost"))).thenReturn(username);
+        when(usersRepository.getUser(new MailAddress(RECEIVER_DOMAIN_COM))).thenReturn(username);
+        when(mailboxManager.getMailbox(eq(inbox), any(MailboxSession.class))).thenReturn(messageManager);
+        when(session.getUser()).thenReturn(User.fromUsername(username));
```
Comment (chibenwa) [inline server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/delivery/LocalDeliveryTest.java]: This entire class looks like it should be reworked as it massively relies on mocks.

For now I would ignore it - as these tests are really ugly and need a complete rework (not your fault)

File: server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/delivery/ToRecipientFolderTest.java [hunk 19]
```diff
@@ -77,8 +77,7 @@ public void setUp() throws Exception {
         mailboxManager = mock(MailboxManager.class);
 
 
-        MetricFactory metricFactory = mock(MetricFactory.class);
-        when(metricFactory.generate(anyString())).thenReturn(mock(Metric.class));
+        metricFactory = new RecordingMetricFactory();
```
Comment (chibenwa) [inline server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/delivery/ToRecipientFolderTest.java]: Same thing here, these tests relies too much on mocks to be consensual. Metrics were not tested. Using `NoopMetricFactory` here would be a good move.

File: server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/remote/delivery/DeliveryRunnable.java (hunk 16 above)
Comment (hoangdat) [reply server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/remote/delivery/DeliveryRunnable.java]: I want to re-use it in `DeliveryRunnableTest`

**Comment by trantienduchn** [comment]
read it

File: server/queue/queue-activemq/src/test/java/org/apache/james/queue/activemq/ActiveMQMailQueueBlobTest.java (hunk 14 above)
Comment (hoangdat) [reply server/queue/queue-activemq/src/test/java/org/apache/james/queue/activemq/ActiveMQMailQueueBlobTest.java]: add more method for get metric with prefixName in `RecordingMetricFactory`

File: metrics/metrics-tests/src/test/java/org/apache/james/metrics/tests/RecordingMetricFactoryTest.java [hunk 20]
```diff
@@ -0,0 +1,102 @@
...
+        assertThat(testee.countFor(METRIC_NAME))
+            .isEqualTo(5);
+    }
+
+    private void runTimeMetric(java.time.Duration duration) {
+        testee.runPublishingTimerMetric(TIME_METRIC_NAME, () -> {
+            try {
+                Thread.sleep(duration.toMillis());
+            } catch (InterruptedException e) {
+                e.printStackTrace();
```
Comment (chibenwa) [inline metrics/metrics-tests/src/test/java/org/apache/james/metrics/tests/RecordingMetricFactoryTest.java]: Please throw a runtime here

File: metrics/metrics-tests/src/test/java/org/apache/james/metrics/tests/RecordingMetricFactoryTest.java [hunk 21]
```diff
@@ -0,0 +1,102 @@
...
+    void executionTimesForATimeMetricShouldBeStoreMultipleTime() {
+        runTimeMetric(ONE_SECOND);
+        runTimeMetric(FIVE_SECONDS);
+
+        WAIT_CONDITION
+            .until(() -> {
+                assertThat(testee.executionTimesFor(TIME_METRIC_NAME))
+                    .hasSize(2);
+                assertThat(testee.executionTimesFor(TIME_METRIC_NAME))
+                    .contains(ONE_SECOND, FIVE_SECONDS);
```
Comment (chibenwa) [inline metrics/metrics-tests/src/test/java/org/apache/james/metrics/tests/RecordingMetricFactoryTest.java]: Please note that strict equality here can not really be enforced.

I'd rather check that `[0]` is at least 1 scond, then check that `[1]` is at least 5 seconds

File: metrics/metrics-tests/src/test/java/org/apache/james/metrics/tests/RecordingMetricFactoryTest.java [hunk 22]
```diff
@@ -0,0 +1,102 @@
...
+    void setUp() {
+        testee = new RecordingMetricFactory();
+    }
+
+    @Test
+    void executionTimesForATimeMetricShouldBeStoreMultipleTime() {
+        runTimeMetric(ONE_SECOND);
+        runTimeMetric(FIVE_SECONDS);
+
+        WAIT_CONDITION
```
Comment (chibenwa) [inline metrics/metrics-tests/src/test/java/org/apache/james/metrics/tests/RecordingMetricFactoryTest.java]: This wait is clearly not needed as we can write this test in a synchronous monothread fashion:

```
TimeMetric timer1 = metricFactory.timer(NAME);
Thead.sleep(ONE_SECOND);
timer1.stopAndPublish();

TimeMetric timer2 = metricFactory.timer(NAME);
Thead.sleep(FIVE_SECOND);
timer2.stopAndPublish();

assertThat(testee.executionTimesFor(TIME_METRIC_NAME)).hasSize(2);
assertThat(testee.executionTimesFor(TIME_METRIC_NAME)).at(0).isGreaterThat(ONE_SECOND);
assertThat(testee.executionTimesFor(TIME_METRIC_NAME)).at(1).isGreaterThat(FIVE_SECOND);
```

**Comment by hoangdat** [comment]
```
[7e248c77858af35df9994c633f6eb8fa991ef91b] [ERROR] Failures: 
[7e248c77858af35df9994c633f6eb8fa991ef91b] [ERROR]   MetricableBlobStoreContract.readBytesShouldPublishReadBytesTimerMetrics:83 
[7e248c77858af35df9994c633f6eb8fa991ef91b] Expected size:<2> but was:<1> in:
[7e248c77858af35df9994c633f6eb8fa991ef91b] <[PT0.001S]>
[7e248c77858af35df9994c633f6eb8fa991ef91b] [ERROR]   MetricableBlobStoreContract.readShouldPublishReadTimerMetrics:93 
```

File: server/queue/queue-api/src/test/java/org/apache/james/queue/api/MailQueueMetricContract.java [hunk 23]
```diff
@@ -66,59 +75,67 @@ default void deQueueMail(Integer times) {
...
+            .containsExactly(2);
     }
 
     @Test
-    default void enqueueShouldPublishEnqueueTimeMetric(MailQueueMetricExtension.MailQueueMetricTestSystem testSystem) throws Exception {
+    default void enqueueShouldPublishEnqueueTimeMetric(MailQueueMetricExtension.MailQueueMetricTestSystem testSystem) {
         enQueueMail(2);
 
-        verify(testSystem.getSpyEnqueuedMailsTimeMetric(), times(2)).stopAndPublish();
+        WAIT_CONDITION
```
Comment (chibenwa) [inline server/queue/queue-api/src/test/java/org/apache/james/queue/api/MailQueueMetricContract.java]: enqueue email is synchronous, so when it returns we should not need to await the metric.

Or did I miss something?

File: metrics/metrics-tests/src/main/java/org/apache/james/metrics/tests/RecordingMetric.java [hunk 24]
```diff
@@ -0,0 +1,57 @@
...
+package org.apache.james.metrics.tests;
+
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Consumer;
+
+import org.apache.james.metrics.api.Metric;
+
+public class RecordingMetric implements Metric {
+    private final String name;
+    private AtomicInteger value;
```
Comment (trantienduchn) [inline metrics/metrics-tests/src/main/java/org/apache/james/metrics/tests/RecordingMetric.java]: final ?

File: metrics/metrics-tests/src/main/java/org/apache/james/metrics/tests/RecordingMetricFactory.java [hunk 25]
```diff
@@ -0,0 +1,79 @@
...
+                .filter(entry -> entry.getKey().startsWith(prefixName))
+                .collect(Multimaps.toMultimap(Map.Entry::getKey, Map.Entry::getValue, HashMultimap::create));
+        }
+    }
+
+    public int countFor(String name) {
+        return executionValues.getOrDefault(name, 0);
+    }
+
+    public List<Integer> countForPrefixName(String prefixName) {
```
Comment (trantienduchn) [inline metrics/metrics-tests/src/main/java/org/apache/james/metrics/tests/RecordingMetricFactory.java]: returning an Integer is better IMHO, because it's a count :)

File: metrics/metrics-tests/src/test/java/org/apache/james/metrics/tests/RecordingMetricFactoryTest.java [hunk 26]
```diff
@@ -0,0 +1,102 @@
...
+        metric.increment();
+        metric.increment();
+        metric.add(3);
+
+        assertThat(testee.countFor(METRIC_NAME))
+            .isEqualTo(5);
+    }
+
+    private void runTimeMetric(java.time.Duration duration) {
+        testee.runPublishingTimerMetric(TIME_METRIC_NAME, () -> {
```
Comment (trantienduchn) [inline metrics/metrics-tests/src/test/java/org/apache/james/metrics/tests/RecordingMetricFactoryTest.java]: how about:
```java
testee.runPublishingTimerMetric(TIME_METRIC_NAME, Throwing.runable(() -> Thread.sleep(...)));
```

File: metrics/metrics-tests/src/main/java/org/apache/james/metrics/tests/RecordingMetricFactory.java (hunk 25 above)
Comment (hoangdat) [reply metrics/metrics-tests/src/main/java/org/apache/james/metrics/tests/RecordingMetricFactory.java]: but if we have more metrics with same prefixName?

Comment (chibenwa) [reply metrics/metrics-tests/src/main/java/org/apache/james/metrics/tests/RecordingMetricFactory.java]: Then maybe a map of matching entries?

File: metrics/metrics-tests/src/main/java/org/apache/james/metrics/tests/RecordingMetricFactory.java [hunk 27]
```diff
@@ -0,0 +1,61 @@
...
+import org.apache.james.metrics.api.MetricFactory;
+import org.apache.james.metrics.api.TimeMetric;
+
+import com.google.common.collect.HashMultimap;
+import com.google.common.collect.Multimap;
+import com.google.common.collect.Multimaps;
+
+public class RecordingMetricFactory implements MetricFactory {
+    private final Multimap<String, Duration> executionTimesInMs = Multimaps.synchronizedSetMultimap(HashMultimap.create());
+    private final ConcurrentHashMap<String, Integer> executionValues = new ConcurrentHashMap<>();
```
Comment (mbaechler) [inline metrics/metrics-tests/src/main/java/org/apache/james/metrics/tests/RecordingMetricFactory.java]: could you name it `counters`?

File: metrics/metrics-tests/src/test/java/org/apache/james/metrics/tests/RecordingMetricFactoryTest.java (hunk 22 above)
Comment (mbaechler) [inline metrics/metrics-tests/src/test/java/org/apache/james/metrics/tests/RecordingMetricFactoryTest.java]: could you name it `WAIT_ONE_MINUTE`?

File: server/queue/queue-api/src/test/java/org/apache/james/queue/api/MailQueueMetricContract.java [hunk 28]
```diff
@@ -30,18 +35,22 @@
...
 import com.github.fge.lambdas.Throwing;
+import com.jayway.awaitility.Duration;
+import com.jayway.awaitility.core.ConditionFactory;
+
 import reactor.core.publisher.Flux;
 import reactor.core.publisher.Mono;
 
 @ExtendWith(MailQueueMetricExtension.class)
 public interface MailQueueMetricContract extends MailQueueContract {
+    ConditionFactory WAIT_CONDITION = await().timeout(Duration.FIVE_SECONDS);
```
Comment (mbaechler) [inline server/queue/queue-api/src/test/java/org/apache/james/queue/api/MailQueueMetricContract.java]: could you name it `WAIT_FIVE_SECONDS`?

**Comment by mbaechler** [comment]
Thank you for this PR, it's so much better than before!

File: metrics/metrics-tests/src/test/java/org/apache/james/metrics/tests/RecordingMetricFactoryTest.java (hunk 22 above)
Comment (hoangdat) [reply metrics/metrics-tests/src/test/java/org/apache/james/metrics/tests/RecordingMetricFactoryTest.java]: It is removed cause of unnecessary wait

File: server/queue/queue-api/src/test/java/org/apache/james/queue/api/MailQueueMetricContract.java (hunk 28 above)
Comment (hoangdat) [reply server/queue/queue-api/src/test/java/org/apache/james/queue/api/MailQueueMetricContract.java]: It is removed cause of unnecessary wait

**Comment by chibenwa** [comment]
` SpamAssassinContract.deletingSpamMessageShouldNotImpactSpamAssassinState:557 » ConditionTimeout`

test this pl

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
```
[65a608ecd30e98d3d2fc0fec5eac651fea77d664] [ERROR]   ReindexingWithEventDeadLettersTest.indexationShouldBeFailingWhenElasticSearchContainerIsPaused:128 » JsonPath
[65a608ecd30e98d3d2fc0fec5eac651fea77d664] [ERROR]   ReindexingWithEventDeadLettersTest.lambda$redeliverShouldCleanEventDeadLetter$6:149 » JsonPath
```

test this please

**Comment by chibenwa** [comment]
Merged
